pageserver/src/tenant.rs

```diff
@@ -552,14 +596,120 @@ impl Tenant {
                 timeline_id,
                 tenant_id,
                 up_to_date_metadata,
-                first_save,
+                false,
             )
             .context("save_metadata")?;
         }
 
         Ok(())
     }
 
+    /// Migrate legacy attaching markers to new location.
+    /// Call this during pageserver startup before loading tenants into memory.
+    /// This code is obsolete on systems where it has run to completion with result `Ok(())`.
+    #[instrument(skip_all)]
+    pub fn migrate_attaching_marker_files(tenants_dir: &Path) -> anyhow::Result<()> {
+        info!("starting on {tenants_dir:?}");
+        let mut dir_entries = std::fs::read_dir(tenants_dir)
+            .with_context(|| format!("list tenants dir {tenants_dir:?}"))?;
+
+        // key: dst, value: src,
+        let mut needed_renames: HashMap<PathBuf, PathBuf> = HashMap::new();
+        loop {
+            let dir_entry = match dir_entries.next() {
+                None => break,
+                Some(res) => res.context("reat tenants dir entry")?,
+            };
+            if !dir_entry.path().is_dir() {
+                continue;
+            }
+
+            let legacy_attach_marker_path = dir_entry
+                .path()
+                .join(TENANT_ATTACHING_LEGACY_MARKER_FILENAME)
+                .to_owned();
+            if !legacy_attach_marker_path.exists() {
+                continue;
+            }
+
+            let legacy_marker_parent = legacy_attach_marker_path.parent().unwrap();
+
+            match legacy_marker_parent
+                .file_name()
+                .and_then(OsStr::to_str)
+                .unwrap()
+                .parse::<TenantId>()
```

Comment:
This unwrap panicking means that we have a non-utf8 path amongst our tenant dirs. I guess I'd prefer to have expect but unsure about the message I'd suggest.

Since this feels so repeated, I think it'd be important to just parse OsStr to an Id but because of windows that is not so easy either.

Well, just noting this. I guess a panic hit here would be easy to resolve by moving the offending dir out. In general it'd be nice to write this as:

```suggestion
            match legacy_marker_parent
                .file_name()
                .and_then(OsStr::to_str)
                .map(|s| s.parse::<TenantId>().map_err(anyhow::Error::new))
                .ok_or_else(|| anyhow::anyhow!("non utf8 path"))
```

And match instead on `Some(was_utf8_convertable_path)` or it wasn't `None`.

Reply:
So, this was basically copy-pasta from `fn init_tenant_mgr`.

And there's a similar pattern in `Tenant::load` for parsing timeline dir names.

Created https://github.com/neondatabase/neon/issues/3488

Reply:
Sounds good.